smolyay/basis.py

```diff
@@ -328,141 +278,138 @@ def derivative(self, x):
         Returns
         -------
         float
-            Value of the derivative of Chebyshev polynomials of first kind.
+            Value of the derivative of Chebyshev polynomials of second kind.
 
         Raises
         ------
         ValueError
-            if input is outside the domain [-1,1].
+            if input is outside the domain [-1, 1].
         """
-        if numpy.any(numpy.greater(x, 1)) or numpy.any(numpy.less(x, -1)):
-            raise ValueError("Input is outside the domain [-1,1]")
-        if self._derivative_polynomial is None:
-            self._derivative_polynomial = ChebyshevFirstKind(self._n+1)
         x = numpy.asarray(x)
         y = numpy.zeros(x.shape)
-        flag2 = x == 1
-        flag3 = x == -1
-        flag1 = ~(flag2 | flag3)
-        y[flag1] =  ((self._n+1)*self._derivative_polynomial(x[flag1]) -
-                     x[flag1]*self(x[flag1]))/(x[flag1]**2-1)
-        y[flag2] = self._n*(self._n + 1)*(self._n + 2)/3
-        y[flag3] = ((-1)**(self._n+1))*self._n*(self._n + 1)*(self._n + 2)/3
+        flag2 = numpy.abs(x) == 1
+        flag1 = ~(flag2)
+        y[flag1] = (
+            (self.n + 1) * scipy.special.eval_chebyt(self.n + 1, x[flag1])
+            - x[flag1] * self(x[flag1])
+        ) / (x[flag1] ** 2 - 1)
+        y[flag2] = (
+            (numpy.sign(x[flag2]) ** (self.n + 1))
+            * self.n
+            * (self.n + 1)
+            * (self.n + 2)
+            / 3
+        )
```

Comment:
I think you've sacrificed a lot of readability here. I would do like this:
```suggestion
        u_limit = self.n * (self.n + 1) * (self.n + 2) / 3
        flag_upper = x == 1
        y[flag_upper] = u_limit
        
        flag_lower = x == -1
        y[flag_lower] = (-1) ** (self.n+1) * u_limit

        flag = ~(flag_upper | flag_lower)
        y[flag] = (
            (self.n + 1) * scipy.special.eval_chebyt(self.n + 1, x[flag])
            - x[flag] * scipy.special.eval_chebyu(self.n, x[flag])
        ) / (x[flag] ** 2 - 1)
```